compiler/src/iree/compiler/Preprocessing/Common/ConvertConvToChannelsLast.cpp

```diff
@@ -0,0 +1,886 @@
+  return success();
+}
+
+namespace {
+
+//=====================================================================
+// Convolution packing patterns
+//=====================================================================
+
+// Named op -> named op conversions if a default inner tile size is specified.
+
+struct ConvertLinalgConvNchwFchw : OpRewritePattern<linalg::Conv2DNchwFchwOp> {
+  using OpRewritePattern::OpRewritePattern;
+  ConvertLinalgConvNchwFchw(MLIRContext *context, PatternBenefit benefit = 2)
+      : OpRewritePattern<linalg::Conv2DNchwFchwOp>(context, benefit) {}
+
+  LogicalResult matchAndRewrite(linalg::Conv2DNchwFchwOp convOp,
+                                PatternRewriter &rewriter) const override {
+    return transposeConvLikeLinalgOp(
+        rewriter, convOp, /*tilingFactor=*/-1,
+        namedConvBuilderFn<linalg::Conv2DNchwFchwOp, linalg::Conv2DNhwcHwcfOp>);
+  }
+};
+
+// Default convolution-like transposing pattern for any linalg op to a generic.
+// Has a lower benefit than the named op conversions.
+
+struct ConvertLinalgConvOp : OpInterfaceRewritePattern<linalg::LinalgOp> {
+  using OpInterfaceRewritePattern<linalg::LinalgOp>::OpInterfaceRewritePattern;
+  ConvertLinalgConvOp(MLIRContext *context, int tile,
+                      PatternBenefit benefit = 1)
+      : OpInterfaceRewritePattern<linalg::LinalgOp>(context, benefit),
+        tilingFactor(tile) {}
+
+  LogicalResult matchAndRewrite(linalg::LinalgOp op,
+                                PatternRewriter &rewriter) const override {
+    return transposeConvLikeLinalgOp(rewriter, op, tilingFactor);
+  }
+
+private:
+  int tilingFactor;
+};
+
+//=====================================================================
+// Generalization patterns
+//=====================================================================
+
+// Returns a reassocation map of the given |rank| such that all dimensions
+// in the set |innerDims| map to two dimensions. For example,
+//
+// rank = 4, innerDims = {1, 3}
+// map = [[0], [1, 2], [3], [4, 5]]
+template <typename SetTy>
+static SmallVector<ReassociationIndices>
+getTilingReassociationMap(int64_t rank, SetTy innerDims) {
+  SmallVector<ReassociationIndices> map;
+  int64_t nTiled = 0;
+  for (int64_t i = 0, e = rank; i < e; i++) {
+    if (innerDims.contains(i)) {
+      map.push_back({i + nTiled++, i + nTiled});
+      continue;
+    }
+    map.push_back({i + nTiled});
+  }
+  return map;
+}
+
+// Generalizes pack operations such that all tiled dimensions have unit outer
+// dims. Produces a transpose on the tiled dimensions followed by an
+// expand_shape to introduce the outer unit dims. For example,
+//
+// tensor.pack inner_dims_pos = [1] inner_tiles = [64]
+//   : tensor<32x64x16xf32> to tensor<32x1x16x64xf32>
+//
+// Generalizes to:
+//
+// linalg.transpose ... tensor<32x64x16xf32> to tensor<32x16x64xf32>
+// tensor.expand_shape ... tensor<32x16x64xf32> to tensor<32x1x16x64xf32>
+class GeneralizeOuterUnitDimsPackOp final
```

Comment:
Do we need the generation patterns. The upstream generalization passes should effectively do this already?

Reply:
The upstream generalization pattern does the expand_shape before the transpose (and also uses insert/extract slices) so unit dim folding doesn't work as well with this. Ideally we could drop this pattern and unit dim folding would handle everything nicely, but for the transient version of this pass we happen to need this pattern right now.

I'll drop the other generalization patterns for the tiling mode though because I don't think we'll use that path right now, and in a final version of the functionality implemented here (via set_encoding) we'd keep the packs around anyway. I only had them because the backends I was targeting didn't support pack.